### PR TITLE
Revert "[Unity][Parser] Check well-formedness in the parser"

### PR DESCRIPTION
**Pull request by tqchen**
Reverts apache/tvm#16569

### CODE REVIEW

**Comment by slyubomirsky** [comment]
Could we perhaps just revert the driver changes if that's what's causing the MLC failure? They are unrelated to the main functionality of the PR and I would rather have the checking for future tests

**Review by Lunderberg — CHANGES_REQUESTED** [review]
@tqchen Can this PR also add a test case for the behavior that should be preserved?  From your comment [here](https://github.com/apache/tvm/pull/16569#issuecomment-2015458566), MLC is currently relying on untested functionality in TVM, and I don't think we should revert based solely on untested functionality.  Doing so would mean that there would neither be a path forward for re-applying the breaking change, nor any protection for the same breakage occurring in other PRs

I think it makes sense to revert commits that break important workflows, so long as there's a reproducible example of the failure.  Otherwise, the responsibility for reproducing that failure mode falls on a developer who wasn't actually seeing the failure.

Edit: Not requesting a perfect unit test that exercises the specific bug, just a high-level test that would trigger the bug in CI.

**Comment by tqchen** [comment]
@slyubomirsky reverting the driver change also works, just wonder if that will break some of the existing tests

**Comment by slyubomirsky** [comment]
The test should have already been broken, oddly! The error was not triggered by anything I added.

**Comment by tqchen** [comment]
agree that we should have a carved out testcases here, will followup in comment soon, as well as in the original thread

I believe it was along the original lines of comment in the changed order, which i assume we say that  we cannot merge dyn smem beyond the kernel boundary, 

```c++
  // MergeSharedMemoryAllocations must be applied after SplitHostDevice
  // because the merged allocation site is at the beginning of each device function
  mixed_pass_list.push_back(tir::transform::MergeSharedMemoryAllocations());
```

**Comment by slyubomirsky** [comment]
`tests/python/tir-transform/test_tir_transform_inject_ptx_async_copy.py::test_vectorize_cp_async_in_if_then_else` was the test that prompted me to change the ordering and it had been failing for several weeks before my PR, oddly. For some reason it was not triggering in CI though.

**Comment by tqchen** [comment]
OK, given that case we should perhaps first revert the driver order only and disable(then look into the failed tests). 

We will still get a UT for the ordering case, which I assume would might relates to getting that dlight function compiled, and then reduce to a smaller case

File: src/tir/transforms/default_gpu_schedule.cc [hunk 1]
```diff
@@ -113,8 +113,7 @@ bool IsScheduledOnGPU(const BaseFunc& func) {
 
   if (target.defined()) {
     int dev_type = target->GetTargetDeviceType();
-    if (!(dev_type == kDLCUDA || dev_type == kDLMetal || dev_type == kDLROCM ||
-          dev_type == kDLWebGPU)) {
+    if (dev_type != kDLCUDA) {
```
Comment (tqchen) [inline src/tir/transforms/default_gpu_schedule.cc:116]: Is this intended change?

Comment (slyubomirsky) [reply src/tir/transforms/default_gpu_schedule.cc:116]: Yes, this was causing a test failure. I will find the link to the discussion

Comment (slyubomirsky) [reply src/tir/transforms/default_gpu_schedule.cc:116]: https://github.com/apache/tvm/pull/16634#issuecomment-1979832554

File: src/tir/ir/data_type_rewriter.cc [hunk 2]
```diff
@@ -532,12 +532,6 @@ Stmt IndexDataTypeRewriter::VisitStmt_(const ForNode* op) {
     n->loop_var = new_loop_var;
     n->min = cast(new_loop_var.dtype(), min);
     n->extent = cast(new_loop_var.dtype(), extent);
-    if (op->thread_binding.defined()) {
-      auto old_thread_binding = op->thread_binding.value();
-      auto* ptr = old_thread_binding.CopyOnWrite();
-      ptr->var = old_thread_binding->var.copy_with_dtype(new_loop_var.dtype());
-      n->thread_binding = std::move(Optional<IterVar>(std::move(old_thread_binding)));
-    }
```
Comment (slyubomirsky) [inline src/tir/ir/data_type_rewriter.cc:540]: This was to address https://github.com/apache/tvm/pull/16634#issuecomment-1973891325

**Comment by slyubomirsky** [comment]
https://github.com/apache/tvm/pull/16770 Smaller-scale reversion

**Comment by Lunderberg** [comment]
Thank you, @tqchen ,  for the follow-up, and for helping to make the more targeted reversion!

**Comment by tqchen** [comment]
closing in favor of https://github.com/apache/tvm/pull/16770

**Pull request closed without merging**